registerWifiNetwork:  saving with WiifiManager.addNetwork (API<=29) or ACTION_WIFI_ADD_NETWORKS (API>=30)

fedepell opened the pull request:
Closes #98, see discussion over there.

daadu commented on android/src/main/java/com/alternadom/wifiiot/WifiIotPlugin.java:
```diff
@@ -616,6 +621,59 @@ public void run() {
+            }
+
+            poResult.success(null);
+        } else {
+            poResult.error("Error", "WifiNetworkSuggestion is not supported for Android SDK " + Build.VERSION.SDK_INT, "");
+        }
+    }
+
```
This code in `connectToDeprecated` is the actual code for "registering network" (it saves the network in android system, also making a connection to it).


The following code in the method does this:
```java
 int updateNetwork = -1;

        if (mWifiConfigList != null) {
            for (android.net.wifi.WifiConfiguration wifiConfig : mWifiConfigList) {
                if (wifiConfig.SSID.equals(conf.SSID)) {
                    conf.networkId = wifiConfig.networkId;
                    updateNetwork = moWiFi.updateNetwork(conf);
                }
            }
        }

        /// If network not already in configured networks add new network
        if (updateNetwork == -1) {
            updateNetwork = moWiFi.addNetwork(conf);
            moWiFi.saveConfiguration();
        }

        if (updateNetwork == -1) {
            return false;
        }
```

We should handle the <Q case for this method by calling it, without checking if the network is connected or not (what is done in `connectToDeprecated`).

My suggestion is strip out this code (with proper return of result) to `registerWifiNetworkDeprecated` - call this in `connectToDeprecated` and `registerWifiNetwork`.

daadu commented on lib/wifi_iot.dart:
```diff
@@ -220,6 +220,26 @@ class WiFiForIoTPlugin {
+    if (!Platform.isIOS && !await isEnabled()) await setEnabled(true);
+    bool bResult;
+    try {
+      await _channel.invokeMethod('registerWifiNetwork', {
```
```suggestion
      bResult = await _channel.invokeMethod('registerWifiNetwork', {
```

daadu commented:
@fedepell `registerWifiNetwork` should be limited to saving the network to the android phone - for  >Q it should be done with the new `Intent` method and for older using `WifiManaged.addNetwork` or `WifiManager.updateNetwork`.

For `NetworkSuggestion` API - there is a PR #99 - which adds `withInternet` flag to `connect` method.

The following code in `connectToDeprecated` method, add/udpate the network to the android in old method. My suggestion is stripping out the code to `registerWifiNetworkDeprecated` (with proper result/return value) and calling this in `connectToDeprecated` and `registerWifiNetwork`

https://github.com/alternadom/WiFiFlutter/blob/352e91cfca2092977856a62c29f9c5cd88b2c350/android/src/main/java/com/alternadom/wifiiot/WifiIotPlugin.java#L921-L941

daadu commented:
@fedepell I would also suggest testing/reviewing #99 for `NetworkSuggestionAPI`, we can merge both of them together.

fedepell commented:
@daadu thanks for the review, sounds fine! I will try to do the changes between this afternoon and tomorrow and report here!

fedepell commented:
@daadu so I added (and tested) also the support for pre-30 devices with the deprecated code (by shuffling a bit around, as you suggested plus I had to shuffle also the code that generates the configuration object). I've dropped the "suggested" network as that is managed separately as per your PR (I will give it a look!). As mentioned I tested a bit both with my devices and emulator.

Hope looks good!

daadu reviewed (approved):
LGTM.

daadu commented:
@Niek Can you review it too?

diegotori commented on android/src/main/java/com/alternadom/wifiiot/WifiIotPlugin.java:
```diff
@@ -616,6 +621,57 @@ public void run() {
+                suggestedNet.setWpa2Passphrase(password);
+            } else if (security != null && security.toUpperCase().equals("WEP")) {
+                // WEP is not supported
+                poResult.error("Error", "WEP is not supported for Android SDK " + Build.VERSION.SDK_INT, "");
```
This is missing a `return` statement. Otherwise, this will fall through and call `poResult.success(null);` further down this block.

fedepell replied:
Many thanks @diegotori for catching this! Fixed now found also one more occurrence from previous code (in that case would not return success but keep doing useless operation that would fail otherwise)

fedepell commented:
ping @Niek for review :)

fedepell commented:
Thanks @Niek ! @daadu can we have it now merged as well?

daadu commented:
This has landed in [v0.2.1 on pub.dev](https://pub.dev/packages/wifi_iot/versions/0.2.1), do check it out. Thank you @fedepell  for your contribution to this package.
